modules/shadertools/src/lib/shader-assembly/resolve-modules.ts

```diff
@@ -6,15 +6,34 @@ import type {ShaderModule} from '../shader-module/shader-module';
 import {ShaderModuleInstance} from '../shader-module/shader-module-instance';
 
 /**
- * Instantiate shader modules and esolve any dependencies
+ * Resolve any dependencies and optionally instantiate modules
  */
 export function resolveModules(
   modules: (ShaderModule | ShaderModuleInstance)[]
-): ShaderModuleInstance[] {
-  const instances = ShaderModuleInstance.instantiateModules(modules);
-  return getShaderDependencies(instances);
+): ShaderModuleInstance[];
+export function resolveModules(
+  modules: (ShaderModule | ShaderModuleInstance)[],
+  instantiate: true
+): ShaderModuleInstance[];
+export function resolveModules(modules: ShaderModule[], instantiate: false): ShaderModule[];
+export function resolveModules(
+  modules: ShaderModuleInstance[],
+  instantiate: false
+): ShaderModuleInstance[];
+export function resolveModules(
+  modules: (ShaderModule | ShaderModuleInstance)[],
+  instantiate: boolean = true
+): (ShaderModule | ShaderModuleInstance)[] {
+  return getShaderDependencies(
```

Comment:
A problem here is that if a ShaderModule has been involved in instantiation, all its dependencies have been converted to `ShaderModuleInstances`. So I am not convinced that the result of any repeated calls to `resolveModules()` will actually be a `ShaderModule[]` in spite of this typing effort.

The ShaderModule / ShaderModuleInstance dichotomy seems to generate a lot more trouble than it is worth. 

I note that instantiation actually doesn't do that much, it adds some methods (that could be global functions) and initializes some things for perf:

```
    this.deprecations = this._parseDeprecationDefinitions(deprecations);
    this.injections = normalizeInjections(inject);
    if (uniformPropTypes) {
      this.uniforms = makePropValidators(uniformPropTypes);
    }
}
```

Perhaps we should try to move away from ShaderModuleInstance.

Reply:
I will try to remove `ShaderModuleInstance` so that we can avoid this mess, partial work in #2074 will continue tomorrow

Reply:
Sounds good, it also makes it simpler for situations where we extend `ShaderModule`, like `ShaderPass`. Why have `ShaderModuleInstance`, but not `ShaderPassInstance`? If I `instantiateModules` with `ShaderPass` instances what is returned etc...